rickshaw

```diff
@@ -0,0 +1,212 @@
+        @params = @{ $params_ref };
+    } elsif ($arg eq "dir") {
+        $base_dir = $val;
+    } elsif ($arg eq "endpoint") {
+        push(@endpoints, $val);
+    } elsif ($arg =~ /^test-order$|^tool-group$|^num-samples$|^name$|^email$|^tags$|^desc|/) {
+        $run{$arg} = $val;
+    } else {
+        printf "[ERROR]parameter not valid: %s\n";
+        exit 1;
+    }
+}
+
+# Apply defaults
+foreach my $p (keys %defaults) {
+    if (! exists $run{$p}) {
+        $run{$p} = $defaults{$p};
+    }
+}
+# Confirm we have required params
+defined $bench_config_ref || die "[ERROR]You must use --bench-config=/path/to/benchmark-config.json\n";
+defined $params_ref || die "[ERROR]Your must use --bench-params=/path/to/benchmark-params.json\n";
+
+-e $base_dir || mkdir($base_dir);
+
+# If there are no endpoints, assume 1 endpoint using the 'local' extension
+if (scalar @endpoints == 0) {
+    if (exists $bench_config{'client'} and exists $bench_config{'server'}) {
+        push(@endpoints, "local:client[1],server[1]");
+    } else {
+        push(@endpoints, "local:client[1]");
+    }
+}
+
+print "\nEndpoints that will be used:\n";
+foreach my $endpoint (@endpoints) {
+    printf "%s\n", $endpoint;
+}
+$run{'endpoints'} = \@endpoints;
+
+
+printf "\nendpoint output:\n";
+# Deploy the endpoints so they are ready to run benchmark and tools
+foreach my $endpoint (@endpoints) {
+    (my $type, my $opts) = split(/:/, $endpoint);
+    if (-e "./endpoints/" . $type) {
+        my $cmd = "./endpoints/" . $type . " " . $opts . " \"" . $bench_config_file . "\"";
+        debug_log(sprintf "going to run %s\n", $cmd);
+        # The below 'system' needs to be forked, then wait for all to finish.
+        # The endpoint program should get all clients/servers "ready", that is,
+        # waiting for instructions from roadblock.  The above command needs
+        # info about how to contact roadblock.
+        my $rc = system ($cmd);
+        printf "exit code from endpoint: %d\n", $rc;
+    } else {
+        printf "[ERROR]could not find endpoint ./endpoints/%s\n", $type;
+        exit 1;
+    }
+}
+
+# Build test execution order
+my @tests;
+if ($run{'test-order'} eq 'i-s') { # Run all samples for first iteration, then all samples for next iteration.
+    for (my $iid = 0; $iid < scalar @params; $iid++) {
+        for (my $sid = 0; $sid < $run{'num-samples'}; $sid++) {
+            my %test = ('iter_id' => $iid, 'samp_id' => $sid);
+            push(@tests, \%test);
+        }
+    }
+} elsif ($run{'test-order'} eq 's-i') { # Run all iterations for first sample, then all for next sample, etc.
+    for (my $sid = 0; $sid < $run{'num-samples'}; $sid++) {
+        for (my $iid = 0; $iid < scalar @params; $iid++) {
+            my %test = ('iter_id' => $iid, 'samp_id' => $sid);
+            push(@tests, \%test);
+        }
+    }
+}
+
+# Cycle through the list of tests using roadblock to kick them off
+my @iterations;
+printf "\nRequesting endpoints to start tools\n";
```

Comment:
Looks like I have some work to do....

Reply:
Well, we will need roadblock client included in the workshop images for sure.  For now for "local" endpoint I can probably just manually install all the redis client stuff (and some benchmark) and concentrate on the other stuff for the local endpoint, like processing the script or command-queue and making sure roadblock communication works.  Then hopefully most of that can be re-used for the ssh endpoint, and when workshop is ready, we'll switch to the fs-image roadblock provides (with chroot or something).